plaidml2/edsl/edsl_test.py

```diff
@@ -386,6 +391,205 @@ def test_use_default(self):
+        K = Tensor(LogicalShape(plaidml.DType.FLOAT32, [3]), name='K')
+        N = TensorDim()
+        M = TensorDim()
+        i = TensorIndex()
+        j = TensorIndex()
+        I.bind_dims(N)
+        K.bind_dims(M)
+        O = TensorOutput(5)
+
+        O[i] += (I[(i - j + 1) // 2] * K[j])
+
+        program = Program('this-is-not an identifier', [O])
+        self.assertMultiLineEqual(
+            str(program), '''function (
+  I[I_0],
+  K[K_0]
+) -> (
+  _X0
+) {
+  _X0[x0 : 5] = +(I[1/2 + 1/2*x0 - 1/2*x1] * K[x1]);
+}
+''')
+        outputs = run(program, [(I, np.array([1, 2, 3])), (K, np.array([1, 2, 3]))])
+        self.assertEqual(outputs[0].tolist(), [2, 5, 4, 9, 6])
+
+    def testTileIdentity(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [3]), name='I')
+        program = Program('tile_identity', [I])
+        self.assertMultiLineEqual(str(program), '''function (
+  I[I_0]
+) -> (
+  _X0
+) {
+  _X0 = ident(I);
+}
+''')
+        outputs = run(program, [(I, np.array([(1, 2, 3)]))])
+        self.assertEqual(outputs[0].tolist(), [1, 2, 3])
+
+    @unittest.skip('TODO: convert to EDSL -  exception needs to be thrown')
+    def testAssignmentExceptions(self):
+        A = Tensor(LogicalShape(plaidml.DType.FLOAT32, [5, 1]), name='A')
+        B = Tensor(LogicalShape(plaidml.DType.FLOAT32, [1, 5]), name='B')
+        L, M, N = TensorDims(3)
+        i, j, k = TensorIndexes(3)
+        A.bind_dims(L, M)
+        B.bind_dims(M, N)
+        O = TensorOutput(L, N)
+        O[i, j] = A[i, k] * B[k, j]
+        program = Program('assignment_non_exception', [O])
+        self.assertMultiLineEqual(
+            str(program), '''function (
+  A[A_0, A_1],
+  B[B_0, B_1]
+) -> (
+  _X0
+) {
+  _X0[x0, x2 : 5, 5] = =(A[x0, x1] * B[x1, x2]);
+}
+''')
+        outputs = run(program, [(A, np.array([[1], [2], [3], [4], [5]])),
+                                (B, np.array([1, 2, 3, 4, 5]))])
+        self.assertEqual(outputs[0].tolist(),
+                         [[1., 2., 3., 4., 5.], [2., 4., 6., 8., 10.], [3., 6., 9., 12., 15.],
+                          [4., 8., 12., 16., 20.], [5., 10., 15., 20., 25.]])
+
+        O = TensorOutput(L, N)
+        O[i, j] = B[i, k] * A[k, j]
+        with self.assertRaises(plaidml.Error) as cm:
+            program = Program('assignment_exception', [O])
+        self.assertTrue("illegal assignment aggregation" in str(cm.exception))
+
+    def testTwoOutputs(self):
+        I = Tensor(LogicalShape(plaidml.DType.FLOAT32, [3]), name='I')
+        program1 = Program('two_outputs', [I, I])
+        self.assertMultiLineEqual(
+            str(program1), '''function (
+  I[I_0]
+) -> (
+  _X1,
+  _X0
+) {
+  _X0 = ident(I);
+  _X1 = ident(I);
+}
+''')
+        outputs = run(program1, [(I, np.array([(1, 2, 3)]))])
+        self.assertEqual(outputs[0].tolist(), [1, 2, 3])
+        self.assertEqual(outputs[1].tolist(), [1, 2, 3])
+
+        N = TensorDim(3)
+        O1 = TensorOutput(N)
```

Comment:
There's no need to say `TensorOutput` here, since we're stomping the value of `O1` and `O2` in the next lines anyways.